BLUESPAWN-client/src/util/permissions/permissions.cpp

```diff
@@ -291,6 +332,174 @@ namespace Permissions {
 		return wName;
 	}
 
+	std::vector<LSA_UNICODE_STRING> Owner:: GetPrivileges() {
+		//Ensure policy handle is initialized
+		if (!bPolicyInitialized) {
+			InitializePolicy();
+			if (!bPolicyInitialized) {
+				LOG_ERROR("Error getting owner privliges, couldn't initialize policy handle.");
+				return std::vector<LSA_UNICODE_STRING>{ };
+			}
+		}
+		PLSA_UNICODE_STRING pReceivedPrivs{ nullptr };
+		ULONG uPrivCount{ 0 };
+		std::vector<LSA_UNICODE_STRING> vPrivs{ };
+		HRESULT hr = LsaNtStatusToWinError(LsaEnumerateAccountRights(lPolicyHandle, GetSID(), &pReceivedPrivs, &uPrivCount));
+		AllocationWrapper awReceivedPrivsHandler{ pReceivedPrivs, 0, AllocationWrapper::NET_ALLOC };
+		if (hr != ERROR_SUCCESS && otType != OwnerType::USER) {
+			LOG_ERROR("Error getting owner privileges. (Error: " << GetLastError() << ")");
+			SetLastError(hr);
+			return std::vector<LSA_UNICODE_STRING>{ };
+		}
+		else if (hr == ERROR_SUCCESS) {
+			for (int i = 0; i < uPrivCount; i++) {
+				vPrivs.emplace_back(pReceivedPrivs[i]);
+			}
+		}
+		//Get privileges from groups that a user belongs to
+		if (otType == OwnerType::USER) {
+			PGROUP_USERS_INFO_0 pGroupInfo{ nullptr };
+			DWORD dEntriesRead{ 0 };
+			DWORD dEntriesTotal{ 0 };
+			NET_API_STATUS stat = NetUserGetLocalGroups(wDomainName.c_str(), wName.c_str(), 0, LG_INCLUDE_INDIRECT , reinterpret_cast<LPBYTE *>(&pGroupInfo), MAX_PREFERRED_LENGTH, &dEntriesRead, &dEntriesTotal);
+			AllocationWrapper awGroupInfoHandler{ pGroupInfo, 0, AllocationWrapper::NET_ALLOC };
+			if (stat != NERR_Success) {
+				LOG_ERROR("Error getting user groups. (Net Error: " << stat << ")");
+				return vPrivs;
+			}
+			//Add all privileges from groups to list of user's privileges
+			for (int i = 0; i < dEntriesRead; i++) {
+				Owner oGroup(pGroupInfo[i].grui0_name);
+				hr = LsaNtStatusToWinError(LsaEnumerateAccountRights(lPolicyHandle, oGroup.GetSID(), &pReceivedPrivs, &uPrivCount));
+				awReceivedPrivsHandler = { pReceivedPrivs, 0, AllocationWrapper::NET_ALLOC };
+				if (hr != ERROR_SUCCESS) {
+					LOG_ERROR("Error getting group privileges. (Error: " << GetLastError() << ")");
+				}
+				else if (hr == ERROR_SUCCESS) {
+					for (int i = 0; i < uPrivCount; i++) {
+						vPrivs.emplace_back(pReceivedPrivs[i]);
+					}
+				}
+			}
+		}
+		SetLastError(ERROR_SUCCESS);
+		return vPrivs;
+	}
+
+	bool Owner::HasPrivilege(IN const std::wstring& wPriv) {
+		auto OwnerPrivs = GetPrivileges();
+		return PrivListHasPrivilege(OwnerPrivs, wPriv);
+	}
+
+	bool Owner::PrivListHasPrivilege(IN const std::vector<LSA_UNICODE_STRING>& vPrivList, IN const std::wstring& wPriv) {
+		for (auto iter = vPrivList.begin(); iter != vPrivList.end(); iter++) {
+			if (wPriv.compare(iter->Buffer) == 0) return true;
```

Comment:
Careful here; UNICODE_STRINGS are only null terminated as a courtesy. Better to call std::wstring{ iter->Buffer, iter->Length }; to ensure your bounds are respected.